keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1811 @@
+$ nft add chain ip kube-proxy services
+$ nft add rule ip kube-proxy services ip daddr . ip protocol . th dport vmap @service_ips
+```
+
+or multiple commands to be executed in a single atomic transaction (as
+with `/sbin/iptables-restore`, but more flexible):
+
+```
+$ nft -f - <<EOF
+add table ip kube-proxy { comment "Kubernetes service proxying rules"; }
+add chain ip kube-proxy services
+add rule ip kube-proxy services ip daddr . ip protocol . th dport vmap @service_ips
+EOF
+```
+
+The syntax for the two modes is the same, other than the need to
+escape shell meta characters in the former case.
+
+When reading data from the kernel (`nft list ...`), `nft` outputs the
+data in a nested "object" form:
+
+```
+$ nft list table ip kube-proxy
+table ip kube-proxy {
+  comment "Kubernetes service proxying rules";
+
+  chain services {
+    ip daddr . ip protocol . th dport vmap @service_ips
+  }
+}
+```
+
+(It is possible to pass data to `nft -f` in this form as well, but
+this wouldn't be useful for us, since we would have to pass the entire
+contents of `table ip kube-proxy` rather than just adding, removing,
+and updating the particular rules, sets, etc, that we wanted to
+change.)
+
+`nft` also has a JSON API, which would theoretically be a better
+option for programmatic use than the "plain text" API. Unfortunately,
+the representation of rules in this mode is vastly different from the
+representation of rules in "plain text" mode:
+
+```
+$ nft --json list table ip kube-proxy | jq .
+...
+    {
+      "rule": {
+        "family": "ip",
+        "table": "kube-proxy",
+        "chain": "services",
+        "handle": 19,
+        "expr": [
+          {
+            "vmap": {
+              "key": {
+                "concat": [
+                  {
+                    "payload": {
+                      "protocol": "ip",
+                      "field": "daddr"
+                    }
+                  },
+                  {
+                    "payload": {
+                      "protocol": "ip",
+                      "field": "protocol"
+                    }
+                  },
+                  {
+                    "payload": {
+                      "protocol": "th",
+                      "field": "dport"
+                    }
+                  }
+                ]
+              },
+              "data": "@service_ips"
+            }
+          }
+        ]
+      }
+...
+```
+
+While it's clear how this _particular_ rule would be converted back
+and forth between the two forms, there is no way to be able to map
+_all_ rules back and forth without having separate code for every rule
+type. Furthermore, the JSON syntax of individual rules is poorly
+documented, and essentially all examples on the web (including the
+nftables wiki, random blog posts, etc) use the non-JSON syntax. So if
+we used the JSON syntax in kube-proxy, it would make the code harder
+to understand and to maintain.
```

Comment:
what I'm still confused is if we can unequivocally can now if we can do a diff of `nft -f myrules.txt` is equal to `nft --json list table ip kube-proxy >myrules.json`, or we don't really need to compare the input against the output?

Reply:
You've got a bunch of extra or missing words there and I'm not sure exactly what you're asking, but anyway, there is no point where we need to actually look at the text of a rule that we previously created.

e.g., if we think that the `services` chain might contain out-of-date rules (because we just started up, and the existing chain might have been created by a previous version of kube-proxy), then there's no point in _looking at_ the chain to try to figure that out. We can just do:

```
	add chain ip kube-proxy services
	flush chain ip kube-proxy services
	add rule ip kube-proxy services ip daddr . ip protocol . th dport vmap @service-ips
	add rule ip kube-proxy services fib daddr type local fib oif != "lo" ip protocol . th dport vmap @service-nodeports
```

The first line ensures that the chain exists ("add" is a no-op if it already existed), the second line ensures that it is empty, and the third and fourth lines add the 2 rules that we want. This is executed atomically, so the net effect is to ensure that the chain exists and contains exactly the rules we want it to, regardless of what it contained before. And basically all updates work that way. (The iptables proxy works the same way.)

In some cases, we need to be able to get the names of all currently-existing chains and sets, so that in case there are stale ones, we can delete them. (Again, the iptables proxy does the same thing.) But that just requires parsing out the _names_ of chains, not their contents, and we can do that with the JSON API with no problem.

(For comparison, in https://github.com/containernetworking/plugins, there is code where we _do_ need to see if particular rules exist; in particular, the CNI plugins need to be able to distinguish between "there is a rule for `192.168.0.5` and it was created for container `foo`" and "there is a rule for `192.168.0.5` and it was created for container `bar`". The iptables code does this by adding a `--comment` to the rule, containing a hashed containerID. The nftables version will do the same thing, and again, we _can_ do that with the JSON API, because in nftables the comment on a rule is stored in a separate field from the expression, so we can list the rules with the JSON API and just look at the `"comment"` fields to identify them, and the `"handle"` fields to get a numeric ID to use to delete the rule if we don't want it.)

Reply:
I apologize, I realize now  that I broke the sentence more than 4 times 😄 , let me start over. I was just wondering if we can diff the blob that we write with the one that we read and know the differences, instead of relying on the `nft` binary to do the magic 

> In some cases, we need to be able to get the names of all currently-existing chains and sets, so that in case there are stale ones, we can delete them. (Again, the iptables proxy does the same thing.) But that just requires parsing out the names of chains, not their contents, and we can do that with the JSON API with no problem.

 seems that based on this comment we do, indeed, I was worried about detecting stale chains

Reply:
> I was just wondering if we can diff the blob that we write with the one that we read and know the differences

In general, no, not really, because of syntactical ambiguities. eg, you might write a rule like:

```
    nft add rule ip kube-proxy postrouting  mark and 0x2000 == 0x2000  masquerade
```

but if you read that back you'll see

```
    chain postrouting {
        meta mark & 0x00002000 == 0x00002000 masquerade
    }
````

This is the other reason why not being able to translate rules back from the JSON form isn't a big deal: because even if you had them in the plaintext form, they might look unrecognizably different from what you were expecting.

> indeed, I was worried about detecting stale chains

Yeah, even in iptables, we don't do that by diffing the actual `iptables-save` and `iptables-restore` data against each other anyway; we parse out the chain names from the `iptables-save` output, and then compare that against another set we build up while generating the `iptables-restore` input.